analysis/src/scope_analyzer.rs

```diff
@@ -0,0 +1,91 @@
+use codespan::Span;
+use parser::{Statement, StatementKind};
+use std::collections::HashMap;
+
+#[derive(Clone, Debug)]
+pub struct Scope {
+    pub name: String,
+    pub span: Span,
+    pub children: Vec<Scope>,
+}
+
+impl Scope {
+    fn find_inner_scope(&self, index: usize) -> Option<Vec<Scope>> {
+        if index < self.span.start || index >= self.span.end {
+            return None;
+        }
+        
+        for child in &self.children {
+            if let Some(inner_scope) = child.find_inner_scope(index) {
+                return Some([&[self.clone()], &inner_scope[..]].concat());
+            }
+        }
+        
+        Some(vec![self.clone()])
+    }
+}
+
+pub struct ScopeAnalyzer {
+    pub statements: Vec<Statement>,
+    pub stack: Vec<String>,
+    pub symtab: HashMap<String, Span>,
+}
+
+impl ScopeAnalyzer {
+    pub fn new(statements: Vec<Statement>) -> Self {
+        Self {
+            statements,
+            stack: Vec::new(),
+            symtab: HashMap::new(),
+        }
+    }
+
+    pub fn analyze(&mut self) -> Vec<Scope> {
+        let statements = self.statements.clone();
+        let scopes: Vec<Scope> = statements.iter().cloned().flat_map(|statement| self.parse_statement(statement)).collect();
+
+        scopes
+    }
+
+    fn parse_statement(&mut self, statement: Statement) -> Option<Scope> {
+        match statement.kind {
+            StatementKind::Scope(Some(name), statements) => {
```

Comment:
```suggestion
            StatementKind::Scope(Some(name), statements)
            | StatementKind::Procedure(name, statements) => {
```
Both the `Scope` and `Procedure` arms had the same content so you could combine them into one arm
(not sure if you like the `|` at the beginning of the second line or end of the first)

Reply:
Technically procedures require names while scopes don't. Hence the separation